man/cat1/markstat.1



MARKSTAT(1)              TIPSY COMMANDS               MARKSTAT(1)



NAME
     markstat - print physical parameters of marked particles

SYNOPSIS
     markstat _b_o_x _p_a_r_t_i_c_l_e

ALIASES
     mstat

DESCRIPTION
     markstat prints physical parameters of the marked  particles
     in box _b_o_x-_n_u_m_b_e_r for the marked particles of type _p_a_r_t_i_c_l_e.
     Parameters that are listed are: the mass, the  density,  the
     volume  of  the box, the size of the box, the center of mass
     coordinates, the center of mass velocity,  and  the  angular
     momentum per unit mass in the center of mass frame.  This is
     similar to the boxstat command but only for the marked  par-
     ticles.

     Particles can be marked using the markarray, markbox,  mark-
     gal,  and readmark commands. Marked particles can be written
     to a file using the writemark command.

     Particles in a particular box  can  be  unmarked  using  the
     unmarkbox  command  and  all  the  marked  particles  can be
     unmarked using  the  unmarkall  command.   Particles  remain
     marked  even when a new time frame is loaded but after a new
     time frame is loaded any particles that were marked  in  the
     previous time frame can only be unmarked using the unmarkall
     command.

     If no particles are marked  or  if  box  _b_o_x-_n_u_m_b_e_r  is  not
     loaded an error is generated.

OPTIONS
  Particle
     The possible choices for _p_a_r_t_i_c_l_e are:

          all
             Show the physical parameters of all the marked  par-
             ticles.

          baryon
             Show the physical parameters of only the marked  gas
             and star particles.

          dark
             Show the physical parameters of only the marked dark
             matter particles.

          gas
             Show the physical parameters of only the marked  gas



KQ Release 2.0     Last change: 22 MARCH 1994                   1






MARKSTAT(1)              TIPSY COMMANDS               MARKSTAT(1)



             particles.

          star
             Show the physical parameters of only the marked star
             particles.

SEE ALSO
     boxstat(1),  markarray(1),  markbox(1),  markgal(1),   read-
     mark(1), unmarkall(1), unmarkbox(1), writemark(1)














































KQ Release 2.0     Last change: 22 MARCH 1994                   2





code/ChangeLog
Wed Jan 11 13:09:36 1995  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (SRCS): added alloca.c so it will be included in the
	distribution.

Wed Dec 28 13:42:47 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* readmark.c: close file after it is read.

	* main.c: RedrawViewport(): make can_size the maximum of
	can_height and can_width.

	* canvas_resize_proc.c: Same as above.

Mon Dec  5 09:39:44 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* xray_lum_load.c, star_lum.c, star_lum_redshift.c: Changed
	argument to my_gets() to blank string.

Fri Dec  2 08:21:50 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (READLINE_SRC): delete config.h.
	(tipsy): added dependence on ALLOCA.

	* fits.c: Added test for endianness.

Thu Dec  1 15:14:58 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* vista.c: fixed initialization of quantity[].

	* find_shape.c: Forced inertia_cm to be symmetric.  This avoid
	problems with jacobi().

Tue Nov  1 11:26:56 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in : fits_xray.c is now gone.  fits() is used instead.
	defs.h, xray.c, vista.c, fits.c and arguments.c are also changed.

Mon Oct 31 09:43:29 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* main.c: RedrawColorbar(): fixed off-by-one drawing errors.

Wed Oct 12 14:27:16 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Created version 2.1.2.

	* absorb.c: New function for calculating aborbtion spectra.

Tue Sep  6 16:06:02 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline/: Replaced with readline-2.0 from prep.ai.mit.edu.

	* my_gets.c: Changed cast to match new external from readline/.

Fri Jul 29 10:25:49 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Ensure that the axes projections do not overflow.

Fri Jul 15 17:28:15 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* profile.c: Check to see if fopen() fails.

Wed Jul 13 09:46:18 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h: Decreased INTMAX to 2^29-1.  (One bit for sign, one bit
	to do INTMAX - (-INTMAX), and one bit for zoom out.

Tue Jul 12 15:55:17 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* canvas_resize_proc.c: Call plot_sub() and draw_label() for
	servers that automatically do backing store and don't generate an
	expose event.

	* main.c: ZoomIn(), ZoomOut(), Pan(), SetBox(): Ignore button
	presses in inactive windows.
	
	* main.c, window.c: Changed ResizeRequest to ConfigureNotify.

Sat Jun  4 13:12:06 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* fits.c, and fits_xray.c: Changed write() to fwrite() for
	portability.

Thu May  5 10:53:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Increase clipping range of velocities to -2*INTMAX <
	v < 2*INTMAX.
	* defs.h: Change INTMAX to 2^30-1 to accommodate the above
	clipping.

Tue Apr 19 18:00:14 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* comm.h: NSK: added pcenter and writeboxrot commands.

Mon Apr 18 14:14:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Add "unzoom" button.

Fri Apr 15 16:01:56 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Fixed bug with color of marked particles.
	Deleted extra conversion to screen coordinates.
	
	* maxpos.c: Make the margins a little smaller.

Thu Mar 31 14:03:11 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Changed XDrawArc to XFillArc for vector draws.

Fri Mar  4 18:09:25 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Added Neal's changes for read_vector() and "master".

Mon Feb 28 11:51:39 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* openascii.c: Added code from Peter Teuben to uncompress files.

Thu Feb 24 13:21:50 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Changed current_plot logical to current_project and
	current_color.

Tue Feb 22 11:24:05 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline: Updated to readline from gdb-1.12

Thu Feb 17 11:46:07 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Implimented a software "z-buffer".  Only draw
	particles with the highest particle color.

Wed Feb 16 13:21:35 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Merged plot_all and plot_vel into one routine.  A
	new routine all_color() is created.  Changes were also needed in
	the view_*() routines.

Sat Nov 13 11:35:01 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Fixed typo for increasing ndark.

Wed Nov  3 14:35:56 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Deleted skip of XClearWindow on first time through.

Thu Oct 28 10:23:50 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* hard.c: Added movie parameter.

Tue Oct 26 11:29:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* alloca.c: New file in case system doesn't provide alloca().

	* vista.c: Included <stdlib.h> for malloc() definition.

Mon Oct 25 19:16:52 1993  Tom Quinn  (trq@virgo)

	* main.c: Attempt to find a depth 8 PseudoColor visual.

Thu Oct 21 08:56:32 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Set backgrounds of gc_color[] to black.

Tue Oct 19 10:56:36 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h, main.h, make_macro.c, read_macro.c, write_macro.c
	run_macro.c: re-vamped macro storage using linked lists.

Mon Oct 18 15:11:34 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* draworient.c: Added cast from Arg to long.

Fri Oct 15 10:04:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Added a check for index < 1.

	* plot_sub.c: Added a check for binary_loaded.
	Added check for boxes_loaded[0].

Wed Oct  6 09:13:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* zoom_window.c: Added zoom command.  Also changed comm.h and
	arguments.c.

Tue Oct  5 17:27:43 1993  Tom Quinn  (trq@hermes.astro.washington.edu

code/markstat.c
#include "defs.h"
#include "fdefs.h"
markstat(job)
    char job[MAXCOMM] ;
{
    char command[MAXCOMM] ;
    int box ;
    char type[MAXCOMM] ;
    double mass ;
    double rho ;
    double x_size,y_size,z_size ;
    Real com[MAXDIM] ;
    Real com_vel[MAXDIM] ;
    Real angular_mom[MAXDIM] ;
    Real ang_mom[MAXDIM] ;
    int i ;
    struct gas_particle *gp ;
    struct dark_particle *dp ;
    struct star_particle *sp ;

    if(sscanf(job,"%s %d %s",command,&box,type) == 3) {
	if (strcmp(type,"gas") != 0 && strcmp(type,"all") != 0 && 
		strcmp(type,"dark") != 0 && strcmp(type,"baryon") != 0 &&
		strcmp(type,"star") != 0) {
	    printf("<sorry, %s is not a proper data type, %s>\n",type,title) ;
	    return ;
	}
	if (boxes_loaded[box]) {
	    x_size = sqrt((boxes[box].x4[0]-boxes[box].x2[0]) * 
		    (boxes[box].x4[0]-boxes[box].x2[0]) +
		    (boxes[box].x4[1]-boxes[box].x2[1]) *
		    (boxes[box].x4[1]-boxes[box].x2[1]) +
		    (boxes[box].x4[2]-boxes[box].x2[2]) *
		    (boxes[box].x4[2]-boxes[box].x2[2])) ;
	    y_size = sqrt((boxes[box].x2[0]-boxes[box].x1[0]) * 
		    (boxes[box].x2[0]-boxes[box].x1[0]) +
		    (boxes[box].x2[1]-boxes[box].x1[1]) *
		    (boxes[box].x2[1]-boxes[box].x1[1]) +
		    (boxes[box].x2[2]-boxes[box].x1[2]) *
		    (boxes[box].x2[2]-boxes[box].x1[2])) ;
	    z_size = sqrt((boxes[box].x5[0]-boxes[box].x3[0]) * 
		    (boxes[box].x5[0]-boxes[box].x3[0]) +
		    (boxes[box].x5[1]-boxes[box].x3[1]) *
		    (boxes[box].x5[1]-boxes[box].x3[1]) +
		    (boxes[box].x5[2]-boxes[box].x3[2]) *
		    (boxes[box].x5[2]-boxes[box].x3[2])) ;
	    mass = 0.0 ;
	    for (i = 0 ; i < header.ndim ; i++) {
		com[i] = 0.0 ;
		com_vel[i] = 0.0 ;
		angular_mom[i] = 0.0 ;
	    }
	    if (strcmp(type,"gas") == 0 || strcmp(type,"all") == 0 || 
		    strcmp(type,"baryon") == 0 ) {
		for (i = 0 ;i < boxlist[box].ngas ;i++) {
		    if(mark_gas[boxlist[box].gpi[i]]){
			gp = boxlist[box].gp[i] ;
			mass += gp->mass ;
			add_const_mult_vec(com, gp->mass, gp->pos) ;
			add_const_mult_vec(com_vel, gp->mass, gp->vel) ;
			cross_product(ang_mom, gp->pos, gp->vel) ;
			add_const_mult_vec(angular_mom, gp->mass, ang_mom) ;
		    }
		}
	    }
	    if (strcmp(type,"dark") == 0 || strcmp(type,"all") == 0) {
		for (i = 0 ;i < boxlist[box].ndark ;i++) {
		    if(mark_dark[boxlist[box].dpi[i]]){
			dp = boxlist[box].dp[i] ;
			mass += dp->mass ;
			add_const_mult_vec(com, dp->mass, dp->pos) ;
			add_const_mult_vec(com_vel, dp->mass, dp->vel) ;
			cross_product(ang_mom, dp->pos, dp->vel) ;
			add_const_mult_vec(angular_mom, dp->mass, ang_mom) ;
		    }
		}
	    }
	    if (strcmp(type,"star") == 0 || strcmp(type,"all") == 0 || 
		    strcmp(type,"baryon") == 0 ) {
		for (i = 0 ;i < boxlist[box].nstar ;i++) {
		    if(mark_star[boxlist[box].spi[i]]){
			sp = boxlist[box].sp[i] ;
			mass += sp->mass ;
			add_const_mult_vec(com, sp->mass, sp->pos) ;
			add_const_mult_vec(com_vel, sp->mass, sp->vel) ;
			cross_product(ang_mom, sp->pos, sp->vel) ;
			add_const_mult_vec(angular_mom, sp->mass, ang_mom) ;
		    }
		}
	    }
            if(mass == 0.){
                printf("<sorry, no particles are marked, %s>\n",title) ;
                return ;
            }
	    for (i = 0 ; i < header.ndim ; i++) {
		com[i] /= mass ;
		com_vel[i] /= mass ;
	    }
	    cross_product(ang_mom, com, com_vel) ;
	    add_const_mult_vec(angular_mom,-mass, ang_mom) ;
	    for (i = 0 ; i < header.ndim ; i++) {
		angular_mom[i] /= mass ;
	    }
	    printf("<physical parameters of the marked %s particles in box %d>\n"
		    ,type,box);
	    rho = mass / boxes[box].volume ;
	    printf("mass = %.6e, density = %.6e, volume = %.6e\n",
		    mass, rho, boxes[box].volume) ;
	    printf("size = (%.6e x %.6e x %.6e)\n",x_size,y_size,z_size) ;
	    printf("center of mass coordinates = {%.6e, %.6e, %.6e}\n",
		    com[0],com[1],com[2]) ;
	    printf("center of mass velocity = {%.6e, %.6e, %.6e}\n",
		    com_vel[0],com_vel[1],com_vel[2]) ;
	    printf("angular momentum vector = {%.6e, %.6e, %.6e}\n",
		    angular_mom[0],angular_mom[1],angular_mom[2]) ;
	}
	else {
	    printf("<sorry, box %d is not loaded, %s>\n",box,title) ;
	}
    }
    else {
	input_error(command) ;
    }
}
